Include/Rover/RandomPick.hpp

```diff
@@ -0,0 +1,72 @@
+#ifndef ROVER_RANDOM_PICK_HPP
+#define ROVER_RANDOM_PICK_HPP
+#include "Rover/Evaluator.hpp"
+#include "Rover/Pick.hpp"
+#include "Rover/Range.hpp"
+
+namespace Rover {
+
+  //! An argument generator that evaluates to one of the arguments with an
+  //! equal likelyhood.
+  /*!
+    \tparam G The types of the generators.
+  */
+  template<typename... G>
+  class RandomPick {
+    private:
+      class Choice {
+        public:
+          explicit Choice(std::size_t num_generators);
+          std::size_t operator ()();
+
+        private:
+          Range<std::size_t, std::size_t> m_range;
+      };
+      using PickType = Pick<Choice, G...>;
+
+    public:
+
+      //! The type of generated values.
+      using Type = typename PickType::Type;
+
+      //! Constructs RandomPick.
+      /*!
+        \param generators One or more generators evaluating to the same type.
+      */
+      template<typename... GeneratorsFwd>
+      explicit RandomPick(GeneratorsFwd&&... generators);
+
+      //! Evaluates the generator.
+      Type generate(Evaluator& evaluator);
+
+    private:
+      PickType m_pick;
+  };
+
+  template<typename... GeneratorsFwd>
+  RandomPick(GeneratorsFwd&&...) -> RandomPick<std::decay_t<GeneratorsFwd>...>;
+
+  template<typename... G>
+  RandomPick<G...>::Choice::Choice(std::size_t num_generators)
+    : m_range(0, num_generators - 1) {}
+
+  template<typename... G>
+  std::size_t RandomPick<G...>::Choice::operator ()() {
+    auto evaluator = Evaluator();
```

Comment:
Look into reusing the ```Evaluator``` provided in the ```generate``` method rather than constructing a new one. There are plans to heavily optimize ```Evaluator``` so that ```Evaluator```s can be reused.

Reply:
My motivation was that I did not have to store the `Evaluation `for this `Range` in `Evaluator`, since the execution of this function is tightly bound to `RandomPick::generate`. `Evaluator` currently has a linear lookup complexity (constant in we are to use a hash table instead but, as you mentioned, with a much worse coefficient), so bloating it is not a good idea either. That said, I assumed the overhead of constructing an empty vector in `Evaluator::Evaluator() = default` is better.
Optimization-wise, perhaps it is better if we add a stateless `generate()` method to the `Generator` concept. Otherwise, it is faster to use a free random number generator instead of `Range` anyway.

Reply:
Anyway, once `Choice` is a `Generator`, I should save its value in `Evaluator` so that multiple occurrences of the same `Choice` evaluate to the same value within the same session.